### ISSUE TITLE
use docker credential helper for registry login (config.json)

### ISSUE THREAD

**Issue by tbugfinder** [apptainer/singularity#2792]
## Version of Singularity:

singularity version 3.1.0-rc3

### Expected behavior

In case registry authentication is required singularity should make use of configured credential helpers in ~/.docker/config.json.

### Actual behavior

Singularity ends up with:
```
INFO:    Starting build...
FATAL:   Unable to pull docker://<snip>.amazonaws.com/singularitylibrary/repo/image1:latest: conveyor failed to get: Error reading manifest latest in <snip>.amazonaws.com/singularitylibrary/repo/image1: unauthorized: authentication required

``` 

### Steps to reproduce behavior

- Build & install ecr-credential-helper helper (https://github.com/awslabs/amazon-ecr-credential-helper), including config.json
- Use e.g. Amazon ECR for storing images
- pull an image from there.

**Comment by jscook2345** [comment]
@tbugfinder 

Out of curiosity, did the option `--docker-login` work when pulling?

<pre>
[vagrant@localhost ~]$ singularity help pull
...
--docker-login      interactive prompt for docker authentication
...
</pre>

**Comment by tbugfinder** [comment]
@jscook2345 I was testing usig ENV variables:
SINGULARITY_DOCKER_PASSWORD=XXXXXX #using output of the awscli tool
SINGULARITY_DOCKER_USERNAME=AWS

**Comment by tbugfinder** [comment]
Does anyone know how to get support for that?
Oh didn't see that it was already tagged.

**Comment by jscook2345** [comment]
@tri-adam @bauerm97 @ikaneshiro 

Do any of you have experience pulling from AWS using the docker transport?

**Comment by tbugfinder** [comment]
@jscook2345 @tri-adam @bauerm97 @ikaneshiro
an action or discussion on that issue would be appreciated.
Would a PR pass the gates?

**Comment by ikaneshiro** [comment]
Hey @tbugfinder, sorry this has gotten so stale.

Singular will check `~/.docker/config.json` for credentials, but it appears to only work for credentials generated with `docker login`. So you can work around this for now by using `aws ecr get-login` to generate the `docker login` command you need to properly store your credentials in the config. 

This is more of a workaround than a solution to this issue imo. I think it would be better if singularity could work with the `ecr-credential-helper` in the same manner as `docker`, so I will investigate further to see how we can integrate that as well.

**Comment by tbugfinder** [comment]
Hi @ikaneshiro,
the implementation should also support different registries. The helper `ecr-credential-helper` should be invoked for AWS, but not for plain `docker://ubuntu` images.
If there are different AWS credentials the appropriate credential should be applied.

**Comment by ikaneshiro** [comment]
Hey @tbugfinder,
After doing some digging, it turns out we currently support `credHelpers`, but not `credsStore` in docker configs. That is you can set something like:
```
{
	"credHelpers": {
		"aws_account_id.dkr.ecr.region.amazonaws.com": "ecr-login"
	}
}
```
But not:
```
{
	"credsStore": "ecr-login"
}
```

Is `credHelpers` support sufficient for your use case?

**Comment by tbugfinder** [comment]
Singularity only searches for the helper within a few specific folders although it is reporting to search in $PATH.



[pid 396991] newfstatat(AT_FDCWD, "/bin/docker-credential-ecr-login",  <unfinished ...>
[pid 396910] nanosleep({0, 10000000},  <unfinished ...>
[pid 396991] <... newfstatat resumed> 0xc0003669f8, 0) = -1 ENOENT (No such file or directory)
[pid 396991] newfstatat(AT_FDCWD, "/usr/bin/docker-credential-ecr-login", 0xc000366ac8, 0) = -1 ENOENT (No such file or directory)
[pid 396991] newfstatat(AT_FDCWD, "/sbin/docker-credential-ecr-login", 0xc000366b98, 0) = -1 ENOENT (No such file or directory)
[pid 396991] newfstatat(AT_FDCWD, "/usr/sbin/docker-credential-ecr-login", 0xc000366c68, 0) = -1 ENOENT (No such file or directory)
[pid 396991] newfstatat(AT_FDCWD, "/usr/local/bin/docker-credential-ecr-login", 0xc000366d38, 0) = -1 ENOENT (No such file or directory)
[pid 396991] newfstatat(AT_FDCWD, "/usr/local/sbin/docker-credential-ecr-login", 0xc000366e08, 0) = -1 ENOENT (No such file or directory)

**Comment by tbugfinder** [comment]
After switching to credHelpers I've successfully downloaded an image from ECR.

**Comment by tbugfinder** [comment]
In addition if the credHelper is configured for AWS account A pulling from dockerhub is still possible.

**Comment by ikaneshiro** [comment]
Hey @tbugfinder I'm glad this is working for you. Can you give me some more info around that strace? Did you set a location in your $PATH with the cred helper and singularity was unable to find it because it was only searching "standard" $PATH locations?

**Comment by tbugfinder** [comment]
Hey @ikaneshiro Your understanding is correct.
I've built the helper within default go path and added go/bin to PATH. `docker-credential-ecr-login` is executed from bash-CLI.
However singularity does not execute the binary by evaluating my $PATH (shell, .bashrc), although the error message supposes it (to me).
```

$ which docker-credential-ecr-login
~/go/bin/docker-credential-ecr-login

 $ singularity pull docker://<account>.dkr.ecr.us-west-1.amazonaws.com/test/test2/image:latest
WARNING: Authentication token file not found : Only pulls of public images will succeed
INFO:    Starting build...
FATAL:   Unable to pull docker://<account>.dkr.ecr.us-west-1.amazonaws.com/test/test2/image:latest: conveyor failed to get: error getting username and password: error getting credentials - err: exec: "docker-credential-ecr-login": executable file not found in $PATH, out: `

```

Should $PATH be inherited from the executing shell?

**Comment by ikaneshiro** [comment]
Hey @tbugfinder,
Does it only fail if your credhelper is at a path with a `~` in it? Try using an absolute path. Since the cred-helper is being directly called with exec I don't think there is any shell expansion occurring.

For more info: I have tried to replicate this with credhelpers in non-standard locations and modified the PATH accordingly with no problems. When I tried a path with a tilde I ran into this error though
